### DISCUSSION TITLE
How to recover non-`SpendableOutputs` coins

### DISCUSSION THREAD

**Discussion by luckysori** [lightningdevkit/rust-lightning#2251] [Help] [answered]
We want to be able to recover on-chain funds that have not yet triggered an `SpendableOutputs` event. Is there a simple way to this other than periodically backing up some LDK data directories?

**Comment by wpaulino** [comment, marked as answer]
`SpendableOutputs` events are generated by each channel's `ChannelMonitor`, so as long as you've been persisting and backing up `ChannelMonitor`s and their `ChannelMonitorUpdate`s, you should be good to go.

**Reply by luckysori** [reply]
Thanks, that makes sense! 

We've actually gone in a slightly different direction and instead we are now overriding `get_destination_script` and `get_shutdown_scriptpubkey` from `KeysInterface` (or whatever it's called on the latest version of LDK) so that the coins are always owned by the same key by the time the commitment transaction is confirmed.

**Reply by wpaulino** [reply]
FYI that still requires that you have the `ChannelMonitor` running such that the commitment claims go to those keys. The `get_shutdown_scriptpubkey` only applies for co-op closes, which you wouldn't be able to do with the `ChannelMonitor` data alone.

**Reply by luckysori** [reply]
Okay, I must have been mistaken then. I thought that one could specify these addresses so that the commitment transaction would pay directly to them too, not just collaborative close.

**Reply by wpaulino** [reply]
Not quite. The commitment transactions will always pay to tweaked basepoint keys. The `ChannelMonitor` will then claim the outputs with those same keys, and send the funds to the script returned by `get_destination_script`.

Do you have any suggestions on how we could improve the docs in that area?

**Reply by luckysori** [reply]
Oh, so does that mean we don't actually need to spend the funds of the commitment transaction ourselves when the `SpendableOutputs` event pops if we have overridden `get_destination_script`?

> Do you have any suggestions on how we could improve the docs in that area?

From what I've understood now, I would maybe clarify that `get_shutdown_scriptpubkey` only applies to a _cooperative_ close. I think when I first glanced at the docs and saw `closing` I wrongly assumed it applied to both.

**Reply by wpaulino** [reply]
> Oh, so does that mean we don't actually need to spend the funds of the commitment transaction ourselves when the SpendableOutputs event pops if we have overridden get_destination_script?

You still have to regardless of overriding it or not, but not immediately, since the HTLCs have already been resolved and you no longer run the risk of losing funds.

**Reply by luckysori** [reply]
I'm struggling to understand what you mean by "the HTLCs have already been resolved" in the context of a published commitment transaction.

Are you referring to the fact that the commitment transaction HTLC now pays to the overridden destination script, so we only have to handle one `SpendableOutputs` event rather than two?

**Reply by wpaulino** [reply]
HTLC outputs in a commitment transaction are unresolved because they can only be claimed via pre-agreed upon keys (not from `get_destination_script` or `get_shutdown_scriptpubkey`) with a preimage or after the HTLC expires.

When your commitment transaction confirms, you go through a two-phased approach, to make sure that even if you claim the HTLC via either of these paths, the counterparty still has a chance to punish you if it belonged to a revoked commitment. Once the `ChannelMonitor` claims it via the preimage or timeout, the HTLC is considered resolved, and assuming it's not from a revoked commitment, only you can spend it. The `ChannelMonitor` at this point will give you a `SpendableOutputDescriptor::DelayedPaymentOutput` for you to spend. You can have these funds go anywhere you want, and if you send them to your `get_destination_script`, then you'll get a `SpendableOutputDescriptor::StaticOutput`.

When your counterparty's commitment transaction confirms, it's a bit simpler. The `ChannelMonitor` will send the funds directly to `get_destination_script` if it can claim the HTLC via the preimage or timeout and give you a `SpendableOutputDescriptor::StaticOutput`.

**Reply by luckysori** [reply]
Thanks a lot for the detailed explanation, I think I get it now! I might come back with a few more questions about your last message later if you have time for this, but this is already very helpful.

**Comment by carlostipton1319** [comment]
s a user, if you want to recover on-chain funds that have not triggered a SpendableOutputs event in LDK (Lightning Development Kit), there are a few options to consider:

Check for transaction confirmations: Monitor the blockchain [without disclosing](https://cryptocointricks.com/how-influencers-hype-crypto-without-disclosing/) explorer or use an API to check if the transactions associated with the funds have been confirmed. Once the transactions are confirmed, the funds should become spendable.
